File: patterns/distributed/saga.js
/**
 * Saga Pattern - JavaScript
 * 
 * Manages distributed transactions by breaking them into a series of
 * local transactions with compensating actions.
 */

class SagaStep {
    constructor(name, action, compensation) {
        this.name = name;
        this.action = action;
        this.compensation = compensation;
        this.executed = false;
    }
    
    async execute() {
        try {
            const result = await this.action();
            this.executed = true;
            return result;
        } catch (error) {
            throw error;
        }
    }
    
    async compensate() {
        if (this.executed && this.compensation) {
            await this.compensation();
            this.executed = false;
        }
    }
}

class Saga {
    constructor() {
        this.steps = [];
    }
    
    addStep(step) {
        this.steps.push(step);
    }
    
    async execute() {
        const executedSteps = [];
        
        try {
            for (const step of this.steps) {
                await step.execute();
                executedSteps.push(step);
            }
            return { success: true };
        } catch (error) {
            // Compensate in reverse order
            for (let i = executedSteps.length - 1; i >= 0; i--) {
                await executedSteps[i].compensate();
            }
            return { success: false, error: error.message };
        }
    }
}

// Example usage
console.log('=== Saga Pattern ===\n');

const saga = new Saga();

saga.addStep(new SagaStep(
    'Reserve Inventory',
    async () => {
        console.log('Reserving inventory...');
        return 'Inventory reserved';
    },
    async () => console.log('Releasing inventory...')
));

saga.addStep(new SagaStep(
    'Charge Payment',
    async () => {
        console.log('Charging payment...');
        return 'Payment charged';
    },
    async () => console.log('Refunding payment...')
));

saga.addStep(new SagaStep(
    'Create Order',
    async () => {
        console.log('Creating order...');
        // Simulate failure
        throw new Error('Order creation failed');
    },
    async () => console.log('Cancelling order...')
));

saga.execute().then(result => {
    console.log('Saga result:', result);
});

module.exports = { Saga, SagaStep };

File: patterns/distributed/saga.py
"""
Saga Pattern - Python

Manages distributed transactions by breaking them into a series of
local transactions with compensating actions.
"""

import asyncio


class SagaStep:
    def __init__(self, name, action, compensation):
        self.name = name
        self.action = action
        self.compensation = compensation
        self.executed = False
    
    async def execute(self):
        try:
            result = await self.action()
            self.executed = True
            return result
        except Exception as error:
            raise error
    
    async def compensate(self):
        if self.executed and self.compensation:
            await self.compensation()
            self.executed = False


class Saga:
    def __init__(self):
        self.steps = []
    
    def add_step(self, step):
        self.steps.append(step)
    
    async def execute(self):
        executed_steps = []
        
        try:
            for step in self.steps:
                await step.execute()
                executed_steps.append(step)
            return {'success': True}
        except Exception as error:
            # Compensate in reverse order
            for step in reversed(executed_steps):
                await step.compensate()
            return {'success': False, 'error': str(error)}


# Example usage
if __name__ == "__main__":
    async def main():
        saga = Saga()
        
        saga.add_step(SagaStep(
            'Reserve Inventory',
            lambda: asyncio.sleep(0.1) or print("Reserving inventory..."),
            lambda: asyncio.sleep(0.1) or print("Releasing inventory...")
        ))
        
        saga.add_step(SagaStep(
            'Charge Payment',
            lambda: asyncio.sleep(0.1) or print("Charging payment..."),
            lambda: asyncio.sleep(0.1) or print("Refunding payment...")
        ))
        
        saga.add_step(SagaStep(
            'Create Order',
            lambda: (_ for _ in ()).throw(Exception("Order creation failed")),
            lambda: asyncio.sleep(0.1) or print("Cancelling order...")
        ))
        
        result = await saga.execute()
        print("Saga result:", result)
    
    asyncio.run(main())

